### PR TITLE
Modular tables now clear allowed weapons if +noreplace isn't specified.

### PR DESCRIPTION
**Pull request by MageKing17**
This makes them behave more like ship flags and should bring them around towards how modders intuitively expect using a modular table to override a ship's allowed weapons to work.

Fixes [Mantis 3176](http://scp.indiegames.us/mantis/view.php?id=3176).

ALTERNATIVELY: If others aren't as convinced as I am that this behavior is the intuitive expectation, or somebody points towards released mods that would break, the logic can be reversed and `+noreplace` replaced with `+override` (like weapons.tbl uses for flags) instead.

### CODE REVIEW

File: code/ship/ship.cpp [hunk 1]
```diff
@@ -3291,10 +3302,12 @@ int parse_ship_values(ship_info* sip, const bool is_template, const bool first_t
 		Error(LOCATION, "%s '%s' has too many primary banks (%d).  Maximum for ships is currently %d.\n", info_type_name, sip->name, sip->num_primary_banks, MAX_SHIP_PRIMARY_BANKS);
 	}
 
+	memset(sip->allowed_weapons, 0, sizeof(int) * MAX_WEAPON_TYPES);
```
Comment (asarium) [inline code/ship/ship.cpp]: This should use `sizeof(sip->allowed_weapons)` as the last parameter to make sure future updates to `allowed_weapons` don't break this code.

**Comment by asarium** [comment]
I think this behavior makes the most sense because all other options are overwritten if specified in a modular table.

Most mods probably already expected that modular tables overwrite previous values so this will most likely not break any mod. If it does then the other variant can be implemented.

**Comment by Goober5000** [comment]
Hmm.  I can see arguments for both implementations.  I'd ask @AxemP and Battuta if their mods depend on the current implementation.

File: code/ship/ship.cpp (hunk 1 above)
Comment (MageKing17) [inline code/ship/ship.cpp]: Okay, but changing every instance of `memset`/`memcpy` on an array in ship.cpp to use `sizeof(var)` instead of `sizeof(type) * size_of_array` (for consistency) should probably be a separate PR.

Comment (asarium) [inline code/ship/ship.cpp]: Changing every instance is probably not a good idea because it's likely to introduce bugs but the better version of determining the size of the array should be used whenever possible.

Comment (MageKing17) [inline code/ship/ship.cpp]: My point was more that this is not the only instance of `memset` being called on this specific array and just changing this one instance would be inconsistent, and changing all of them wouldn't really be in the scope of this PR.

Comment (asarium) [inline code/ship/ship.cpp]: I agree that changing all instances is not in the scope of this PR. However, introducing a minor inconsistency here will improve the code quite a bit (IIRC you even mentioned on IRC that you had issues with memsetting too much memory).

Comment (MageKing17) [inline code/ship/ship.cpp]: Yes, although no amount of defensive coding can protect you against accidentally iterating through a for-loop too many times, and it's not memsetting too much memory *now*, is it?

If just this one line gets changed, and then somebody actually does change the size of the `allowed_weapons` array, I think having all-but-one reference breaking to be less useful than just having all of them break. I'd rather have the consistency than introduce better practices piecemeal, especially since it wouldn't be that much effort to just fix all of them at once.

Comment (Goober5000) [inline code/ship/ship.cpp]: I agree with @MageKing17.  I like keeping things consistent, and if someone does decide to change the implementation of `sip->allowed_weapons`, this is one of the most obvious things to look for when refactoring.

Comment (asarium) [inline code/ship/ship.cpp]: Ok, I guess this is alright then.

**Comment by MageKing17** [comment]
I finally got around to checking Blue Planet's bp-shp.tbm entries, and they all copy the whole line and then add to it. Additionally, @AxemP says none of his tables rely on the current behavior. Speaking from personal experience, when I was testing the SSM overhaul, I also copied the whole line and added entries to it, so I think it's pretty safe to say that this is how everyone assumed it ought to work the whole time (and it looks like as soon as that assumption was violated, we got a Mantis issue about it).

**Comment by Goober5000** [comment]
Excellent point about Mantis, and good for following up with Axem.  Okay, I think we can merge this.
